src/bgw/scheduler.c

```diff
@@ -252,6 +255,19 @@ scheduled_bgw_job_transition_state_to(ScheduledBgwJob *sjob, JobState new_state)
 				return;
 			}
 
+			/* If we are unable to reserve a worker go back to the scheduled state */
+			sjob->reserved_worker = ts_bgw_worker_reserve();
+			if (!sjob->reserved_worker)
+			{
+				elog(WARNING,
+					 "failed to launch job %d \"%s\": out of background workers",
+					 sjob->job.fd.id,
+					 NameStr(sjob->job.fd.application_name));
+				scheduled_bgw_job_transition_state_to(sjob, JOB_STATE_SCHEDULED);
```

Comment:
It would be nice if we could put them in a separate queue and the scheduler picks these up after a job completes. That would involve more extensive changes to the scheduler job processing algorithm. So just a thought here.

Reply:
I don't think we actually need a separate queue because we don't change next_start here and thus the logical priority of the job is preserved and it will be picked up first after a job finishes. There is something nice about only having one priority queue.